tests/test_quoting.py

```diff
@@ -166,43 +166,48 @@ def test_unquoting(num, unquoter):
 
 
 @pytest.mark.xfail
-def test_unquoting_badpercent(unquoter):
-    # Test unquoting on bad percent-escapes
-    given = '%xab'
-    expect = given
-    result = unquoter()(given)
-    assert expect == result
+# FIXME: Expected value should be the same as given.
+# See https://url.spec.whatwg.org/#percent-encoded-bytes
+def test_unquoting_bad_percent_escapes_1(unquoter):
+    assert '%' == unquoter()('%')
 
 
 @pytest.mark.xfail
-def test_unquoting_badpercent2(unquoter):
-    given = '%x'
-    expect = given
-    result = unquoter()(given)
-    assert expect == result
+# FIXME: Expected value should be the same as given.
+# See https://url.spec.whatwg.org/#percent-encoded-bytes
+def test_unquoting_bad_percent_escapes_2(unquoter):
+    assert '%x' == unquoter()('%x')
 
 
 @pytest.mark.xfail
-def test_unquoting_badpercent3(unquoter):
-    given = '%'
-    expect = given
-    result = unquoter(given)
-    assert expect == result
+# FIXME: Expected value should be the same as given.
+# See https://url.spec.whatwg.org/#percent-encoded-bytes
+def test_unquoting_bad_percent_escapes_3(unquoter):
+    assert '%xa' == unquoter()('%xa')
 
 
 @pytest.mark.xfail
-def test_unquoting_mixed_case(unquoter):
-    # Test unquoting on mixed-case hex digits in the percent-escapes
-    given = '%Ab%eA'
-    expect = '\u00ab\u00ea'
-    result = unquoter()(given)
-    assert expect == result
+# FIXME: After conversion to bytes, should not cause UTF-8 decode fail.
```

Comment:
@behnam do you remember why you added this specific comment? I don't really know what to make of it…

cc @mjpieters @Dreamsorcerer any ideas what this could mean? This combo of `pytest.raises()` + `xfail` without a clearly stated reason is rather confusing...

Reply:
Isn't it just repeating the note at the top of link?

> It is generally a good idea for sequences of [percent-encoded bytes](https://url.spec.whatwg.org/#percent-encoded-byte) to be such that, when [percent-decoded](https://url.spec.whatwg.org/#string-percent-decode) and then passed to [UTF-8 decode without BOM or fail](https://encoding.spec.whatwg.org/#utf-8-decode-without-bom-or-fail), they do not end up as failure.

Reply:
Had another look. So, the spec says:
> Using anything but [UTF-8 decode without BOM](https://encoding.spec.whatwg.org/#utf-8-decode-without-bom) when input contains bytes that are not [ASCII bytes](https://infra.spec.whatwg.org/#ascii-byte) might be insecure and is not recommended.

Which I think translates to doing `bytes.decode(errors="replace")`. On this commit though, the code was not using replace:
https://github.com/aio-libs/yarl/blob/0e9b288e10e0b41d01fddb3c5cc2c9d8a2d14678/yarl/quoting.py#L132

So, the code must have produced an exception there, when the expected behaviour is that it would produce an output of `�`. I've not checked the code in the latest commit.

Reply:
Although, that exception was actually suppressed and doesn't explain why it was changed to expect a ValueError...

Maybe I'm still misreading the spec, as it refers specifically to the input. So, I think that maybe we should be doing `val.encode(errors="replace")` at the very start of the function, and processing the bytes, rather than the str.

Maybe the comment was related to a similar misreading of the spec..